### PR TITLE
whois: add package

### PR DESCRIPTION
**Pull request by aparcar**
Better a separate package than enabling it via busybox.

Signed-off-by: Paul Spooren <mail@aparcar.org>

Maintainer: me
Compile tested: CI
Run tested: CI

### CODE REVIEW

**Comment by aparcar** [comment]
@neheb help please

**Comment by neheb** [comment]
Needs nls.mk looks like. Probably $(INTL_DEPENDS) as well.

**Comment by aparcar** [comment]
Works, I'd merge this.

**Comment by neheb** [comment]
```
make[3]: Entering directory `/home/mangix/devstuff/openwrt/build_dir/target-powerpc_464fp_musl/whois-5.5.7'
perl make_version_h.pl debian/changelog > version.h
Can't locate autodie.pm in @INC (@INC contains: /usr/local/lib64/perl5 /usr/local/share/perl5 /usr/lib64/perl5/vendor_perl /usr/share/perl5/vendor_perl /usr/lib64/perl5 /usr/share/perl5 .) at make_version_h.pl line 5.
BEGIN failed--compilation aborted at make_version_h.pl line 5.
make[3]: *** [version.h] Error 2
make[3]: *** Deleting file `version.h'
make[3]: Leaving directory `/home/mangix/devstuff/openwrt/build_dir/target-powerpc_464fp_musl/whois-5.5.7'
make[2]: *** [/home/mangix/devstuff/openwrt/build_dir/target-powerpc_464fp_musl/whois-5.5.7/.built] Error 2
make[2]: Leaving directory `/home/mangix/devstuff/openwrt/feeds/packages/utils/whois'
time: package/feeds/packages/whois/compile#0.22#0.21#1.02
make[1]: *** [package/feeds/packages/whois/compile] Error 2
```

Hmmmmmmmm....

**Comment by aparcar** [comment]
Seems like this should be added to the CI?

**Comment by neheb** [comment]
Another issue, if you build libxcrypt before this package, you get

```
mkpasswd.c:328:9: warning: implicit declaration of function 'crypt_gensalt' [-Wimplicit-function-declaration]
  328 |  salt = crypt_gensalt(salt_prefix, rounds, entropy, 64);
      |         ^~~~~~~~~~~~~
mkpasswd.c:328:7: warning: assignment to 'char *' from 'int' makes pointer from integer without a cast [-Wint-conversion]
  328 |  salt = crypt_gensalt(salt_prefix, rounds, entropy, 64);
```

**Comment by neheb** [comment]
Yet another one...

```
Package whois is missing dependencies for the following libraries:
libidn2.so.0
```

**Comment by neheb** [comment]
Fixed version: https://github.com/neheb/packages/commit/a9d2074e705b8fee93eb575e8e32460b34f63918

File: utils/whois/Makefile [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+define Package/whois
+  SECTION:=utils
+  CATEGORY:=Utilities
+  TITLE:=Intelligent WHOIS client
+  DEPENDS:=
+  URL:=https://github.com/rfc1036/whois
+endef
+
+define Package/whois/description
+In 1999 I wrote this Whois client from scratch because the alternatives were
```
Comment (adschm) [inline utils/whois/Makefile:30]: Maybe one should remove the "I" here and rephrase it to passive:
"This Whois client was written from scratch in 1999 because the alternatives were obsolete or bloated."

(The reader will not understand to whom the "I" will correspond in our context here.)

Personally, though, I would just drop this first sentence as the following ones sufficiently explain what's necessary.

Comment (aparcar) [reply utils/whois/Makefile:30]: done

**Comment by neheb** [comment]
I'm having second thoughts about this. It's probably better not to compile mkpasswd given that it's not used.

File: utils/whois/test.sh [hunk 2]
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+
+whois --version | grep "$2"
```
[the remark this answers is not in the document]
Comment (karlp) [reply utils/whois/test.sh:3]: wow, that's a horrific filename, but ok :)

[the remark this answers is not in the document]
Comment (aparcar) [reply utils/whois/test.sh:3]: How would you call it?

**Comment by aparcar** [comment]
> I'm having second thoughts about this. It's probably better not to compile mkpasswd given that it's not used.

It's not installed and I'd like to avoid additional patches, but if you don't like the patch as is I can see how to patch it out.

**Comment by neheb** [comment]
Now I see what the problem is with libxcrypt.

```
#hack to get libxcrypt working
TARGET_LDFLAGS:=-L$(STAGING_DIR)/usr/lib/libxcrypt $(TARGET_LDFLAGS)
```

Needs to be added. musl and libxcrypt both provide a crypt.h include.

Although if you don't plan to provide an mkpasswd package, you can ignore that and just use the current patch.

**Comment by aparcar** [comment]
No more mkpasswd

File: utils/whois/Makefile [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+This client is intelligent and can automatically select the appropriate whois
+server for most queries.
+
+The internal database is often more accurate than IANA's published one, but
+please send me any information you have regarding domains and network resources
+which are not correctly handled by the program.
+endef
+
+EXTRA_CFLAGS+=-DHAVE_GETOPT_LONG
+MAKE_FLAGS += PERL=$(STAGING_DIR_HOST)/bin/perl
```
Comment (neheb) [inline utils/whois/Makefile:39]: This will fail on systems that do not have autodie.pm installed.

I don't know if it's worth adding a build dependency on perl/host.

Comment (aparcar) [reply utils/whois/Makefile:39]: I don't know what that is. If you don't consider it problematic, please merge. If not I'll add perl/host.

**Comment by jefferyto** [comment]
From the README:

> The canonical distribution point for releases of the program is
> http://ftp.debian.org/debian/pool/main/w/whois/ .

File: utils/whois/test.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (karlp) [reply utils/whois/test.sh:3]: ci_autotest.sh perhaps? the discoverability of what the file is for is ~zero

**Comment by aparcar** [comment]
> From the README:

Added

File: utils/whois/test.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (aparcar) [reply utils/whois/test.sh:3]: works for me, I'll create a PR

File: utils/whois/Makefile (hunk 3 above)
Comment (BKPepe) [reply utils/whois/Makefile:41]: So you are upstreaming (to OpenWrt) something which you don't know what it is doing or why it should be there? Wow.

Comment (aparcar) [reply utils/whois/Makefile:41]: @BKPepe I'd appreciate if you adjust your attitude. I'm porting `whois` which happens to contain `mkpasswd`.

> Because of historical reasons this package also contains the mkpasswd
program, which can be used to encrypt a password with crypt(3).

I patched the Makefile so it shouldn't be compiled anymore, which I expected you to figure out on your own.

Comment (aparcar) [reply utils/whois/Makefile:42]: @BKPepe I'd appreciate if you could adjust your attitude. CI builds don't fail with the default packages installed. If there are corner cases where `autodie.pm` is missing, please elaborate.

My local Debian 10 did not require an additional package :shrug:

Comment (aparcar) [reply utils/whois/Makefile]: I adopted the commit based on @neheb feedback here and via IRC, thanks again!
